File: android/src/main/kotlin/ai/nextbillion/kits/nb_kits_flutter/NBKitsFlutterPlugin.kt
package ai.nextbillion.kits.nb_kits_flutter

import ai.nextbillion.kits.geojson.LineString
import ai.nextbillion.kits.geojson.Point
import ai.nextbillion.kits.turf.TurfClassification
import ai.nextbillion.kits.turf.TurfConstants
import ai.nextbillion.kits.turf.TurfConversion
import ai.nextbillion.kits.turf.TurfMeasurement
import ai.nextbillion.kits.turf.TurfMisc
import ai.nextbillion.kits.turf.TurfTransformation
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** NBKitsFlutterPlugin */
class NBKitsFlutterPlugin: FlutterPlugin, MethodCallHandler {
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "nb_kits_flutter")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "nbKits/along" -> {
        val arguments = call.arguments as? Map<*, *>
        arguments?.let { it ->
          val line: List<List<Double>> = it["line"] as List<List<Double>>
          val distance: Double = it["distance"] as Double
          val points = line.map { Point.fromLngLat(it[0], it[1]) }
          val lineString = LineString.fromLngLats(points)
          val along = TurfMeasurement.along(lineString, distance, TurfConstants.UNIT_METERS)
          result.success(along.coordinates())
          return
        }
        result.success(null)

      }

      "nbKits/bearing" -> {
        val arguments = call.arguments as? Map<*, *>
        arguments?.let { it ->
          val point1: List<Double> = it["point1"] as List<Double>
          val point2: List<Double> = it["point2"] as List<Double>
          val bearing = TurfMeasurement.bearing(
            Point.fromLngLat(point1[0], point1[1]),
            Point.fromLngLat(point2[0], point2[1])
          )
          result.success(bearing)
          return
        }
        result.success(null) // Replace with actual result
      }

      "nbKits/circle" -> {
        val arguments = call.arguments as? Map<*, *>
        arguments?.let { it ->
          val center: List<Double> = it["center"] as List<Double>
          val radius: Double = it["radius"] as Double
          val steps: Int = it["steps"] as Int
          val circle = TurfTransformation.circle(
            Point.fromLngLat(center[0], center[1]),
            radius,
            steps,
            TurfConstants.UNIT_METERS
          )
          if (circle.coordinates().isNotEmpty()) {
            val coors = circle.coordinates()[0].map { point ->
              point.coordinates()
            }
            result.success(coors)
            return
          }

        }
        result.success(null)
      }

      "nbKits/degreesToRadians" -> {
        val arguments = call.arguments as? Map<*, *>
        arguments?.let { it ->
          val degrees: Double = it["degrees"] as Double
          val radians = TurfConversion.degreesToRadians(degrees)
          result.success(radians)
          return
        }
        result.success(null)
      }

      "nbKits/destination" -> {
        val arguments = call.arguments as? Map<*, *>
        arguments?.let { it ->
          val point: List<Double> = it["point"] as List<Double>
          val distance: Double = it["distance"] as Double
          val bearing: Double = it["bearing"] as Double
          val destination = TurfMeasurement.destination(
            Point.fromLngLat(point[0], point[1]),
            distance,
            bearing,
            TurfConstants.UNIT_METERS
          )
          result.success(destination.coordinates())
          return
        }
        result.success(null)
      }

      "nbKits/distance" -> {
        val arguments = call.arguments as? Map<*, *>
        arguments?.let { it ->
          val point1: List<Double> = it["point1"] as List<Double>
          val point2: List<Double> = it["point2"] as List<Double>
          val distance = TurfMeasurement.distance(
            Point.fromLngLat(point1[0], point1[1]),
            Point.fromLngLat(point2[0], point2[1]),
            TurfConstants.UNIT_METERS
          )
          result.success(distance)
          return
        }
        result.success(null)
      }

      "nbKits/length" -> {
        val arguments = call.arguments as? Map<*, *>
        arguments?.let { it ->
          val line: List<List<Double>> = it["points"] as List<List<Double>>
          val points = line.map { Point.fromLngLat(it[0], it[1]) }
          val lineString = LineString.fromLngLats(points)
          val length = TurfMeasurement.length(lineString, TurfConstants.UNIT_METERS)
          result.success(length)
          return
        }
        result.success(null)
      }

      "nbKits/lineSlice" -> {
        val arguments = call.arguments as? Map<*, *>
        arguments?.let { it ->
          val line: List<List<Double>> = it["line"] as List<List<Double>>
          val start: List<Double> = it["start"] as List<Double>
          val stop: List<Double> = it["stop"] as List<Double>
          val points = line.map { Point.fromLngLat(it[0], it[1]) }
          val lineString = LineString.fromLngLats(points)
          val lineSlice = TurfMisc.lineSlice(Point.fromLngLat(start[0], start[1]), Point.fromLngLat(stop[0], stop[1]), lineString)
          result.success(lineSlice.coordinates())
          return
        }
        result.success(null)
      }

      "nbKits/lineSliceAlong" -> {
        val arguments = call.arguments as? Map<*, *>
        arguments?.let { it ->
          val line: List<List<Double>> = it["line"] as List<List<Double>>
          val start: Double = it["starDist"] as Double
          val stop: Double = it["stopDist"] as Double
          val points = line.map { Point.fromLngLat(it[0], it[1]) }
          val lineString = LineString.fromLngLats(points)
          val lineSliceAlong = TurfMisc.lineSliceAlong(lineString, start, stop, TurfConstants.UNIT_METERS)
          result.success(lineSliceAlong.coordinates())
          return
        }
        result.success(null)
      }

      "nbKits/nearestPoint" -> {
        val arguments = call.arguments as? Map<*, *>
        arguments?.let { it ->
          val point: List<Double> = it["targetPoint"] as List<Double>
          val points: List<List<Double>> = it["points"] as List<List<Double>>
          val nearestPoint = TurfClassification.nearestPoint(
            Point.fromLngLat(point[0], point[1]),
            points.map { Point.fromLngLat(it[0], it[1]) }
          )
          result.success(nearestPoint.coordinates())
          return
        }
        result.success(null)
      }

      "nbKits/radiansToDegrees" -> {
        val arguments = call.arguments as? Map<*, *>
        arguments?.let { it ->
          val radians: Double = it["radians"] as Double
          val degrees = TurfConversion.radiansToDegrees(radians)
          result.success(degrees)
          return
        }
        result.success(null)
      }

      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

File: ios/Classes/NBKitsFlutterPlugin.swift
import Flutter
import UIKit
import Turf
import CoreLocation

public class NBKitsFlutterPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "nb_kits_flutter", binaryMessenger: registrar.messenger())
        let instance = NBKitsFlutterPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "nbKits/along":
            if let arguments = call.arguments as? [String: Any],
               let line = arguments["line"] as? [[Double]],
               let distance = arguments["distance"] as? Double {
                let points = line.map { CLLocationCoordinate2D(latitude: $0[1], longitude: $0[0]) }
                let lineString = LineString(points)
                if let along = lineString.coordinateFromStart(distance: distance) {
                    result([along.latitude, along.longitude])
                    return
                }
                
            }
            result(nil)
            
        case "nbKits/bearing":
            if let arguments = call.arguments as? [String: Any],
               let point1 = arguments["point1"] as? [Double],
               let point2 = arguments["point2"] as? [Double] {
                let location1 = CLLocationCoordinate2D(latitude: point1[1], longitude: point1[0])
                let location2 = CLLocationCoordinate2D(latitude: point2[1], longitude: point2[0])
                let bearing = location1.direction(to: location2)
                result(bearing)
                return
            }
            result(nil)
            
        case "nbKits/circle":
            if let arguments = call.arguments as? [String: Any],
               let center = arguments["center"] as? [Double],
               let radius = arguments["radius"] as? Double,
               let steps = arguments["steps"] as? Int {
                let circle = circle(center: CLLocationCoordinate2D(latitude: center[1], longitude: center[0]), radius: radius, steps: steps)
                let coordinates = circle.map { point in
                    [point.longitude, point.latitude]
                }
                result(coordinates)
                return
            }
            result(nil)
            
            
        case "nbKits/degreesToRadians":
            if let arguments = call.arguments as? [String: Any],
               let degrees = arguments["degrees"] as? Double {
                let radians = degrees.toRadians()
                result(radians)
                return
            }
            result(nil)
            
        case "nbKits/destination":
            if let arguments = call.arguments as? [String: Any],
               let point = arguments["point"] as? [Double],
               let distance = arguments["distance"] as? Double,
               let bearing = arguments["bearing"] as? Double {
                let location = CLLocationCoordinate2D(latitude: point[1], longitude: point[0])
                let destination = location.coordinate(at: distance, facing: bearing)
                result([destination.longitude, destination.latitude])
                return
            }
            result(nil)
            
        case "nbKits/distance":
            if let arguments = call.arguments as? [String: Any],
               let point1 = arguments["point1"] as? [Double],
               let point2 = arguments["point2"] as? [Double] {
                let location1 = CLLocationCoordinate2D(latitude: point1[1], longitude: point1[0])
                let location2 = CLLocationCoordinate2D(latitude: point2[1], longitude: point2[0])
                let distance = location1.distance(to: location2)
                result(distance)
                return
            }
            result(nil)
            
    
        case "nbKits/length":
            if let arguments = call.arguments as? [String: Any],
               let line = arguments["points"] as? [[Double]] {
                let points = line.map { CLLocationCoordinate2D(latitude: $0[1], longitude: $0[0]) }
                let lineString = LineString(points)
                result(lineString.distance())
                return
            }
            result(nil)
            
            
        case "nbKits/lineSlice":
            if let arguments = call.arguments as? [String: Any],
               let line = arguments["line"] as? [[Double]],
               let start = arguments["start"] as? [Double],
               let stop = arguments["stop"] as? [Double] {
                let points = line.map { CLLocationCoordinate2D(latitude: $0[1], longitude: $0[0]) }
                let lineString = LineString(points)
                let startLocation = CLLocationCoordinate2D(latitude: start[1], longitude: start[0])
                let stopLocation = CLLocationCoordinate2D(latitude: stop[1], longitude: stop[0])
                let coordinates = lineString.sliced(from: startLocation, to: stopLocation)
                let sliced = coordinates.coordinates.map { [$0.longitude, $0.latitude] }
                result(sliced)
                return
            }
            result(nil)
            
        case "nbKits/lineSliceAlong":
            if let arguments = call.arguments as? [String: Any],
               let line = arguments["line"] as? [[Double]],
               let start = arguments["starDist"] as? Double,
               let stop = arguments["stopDist"] as? Double {
                let points = line.map { CLLocationCoordinate2D(latitude: $0[1], longitude: $0[0]) }
                let lineString = LineString(points)
                if let startPoint = lineString.coordinateFromStart(distance: start) {
                    let distance = lineString.distance() - start - stop
                    let coordinates = lineString.trimmed(from: startPoint, distance: distance)
                    let trimmed = coordinates.coordinates.map { [$0.longitude, $0.latitude] }
                    result(trimmed)
                }
                return
            }
            result(nil)
            
            
        case "nbKits/nearestPoint":
            if let arguments = call.arguments as? [String: Any],
               let point = arguments["targetPoint"] as? [Double],
               let points = arguments["points"] as? [[Double]] {
                let location = CLLocationCoordinate2D(latitude: point[1], longitude: point[0])
                let lineString = LineString(points.map { CLLocationCoordinate2D(latitude: $0[1], longitude: $0[0]) })
                if let nearestPoint = lineString.closestCoordinate(to: location) {
                    result([nearestPoint.coordinate.longitude, nearestPoint.coordinate.latitude])
                }
                return
            }
            result(nil)
            
        case "nbKits/radiansToDegrees":
            if let arguments = call.arguments as? [String: Any],
               let radians = arguments["radians"] as? Double {
                let degree = radians.toDegrees()
                result(degree)
                return
            }
            result(nil)
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    func circle(center: CLLocationCoordinate2D, radius: Double, steps: Int) -> [CLLocationCoordinate2D] {
        var coordinates: [CLLocationCoordinate2D] = []
        for i in 0..<steps {
            let angle = Double(i) * 360.0 / Double(steps)
            let destination = center.coordinate(at: radius, facing: angle)
            coordinates.append(destination)
        }
        
        if coordinates.count > 0 {
            coordinates.append(coordinates[0])
        }
        return coordinates
    }
    
}
